### PR TITLE
Ananyamukh6/total variation

### PR DESCRIPTION
**Pull request by ananyamukh6**
Implements total variation

Also adds an example using the added total variation function to denoise an image

### CODE REVIEW

File: examples/total_variation_denoising.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Denoise image using total variation
+======================================
+
+"""
+
+import torch
+import kornia
+import cv2
+import numpy as np
+from torch.autograd import Variable
```
Comment (ducha-aiki) [inline examples/total_variation_denoising.py]: Seems not used here

File: examples/total_variation_denoising.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+# read the image with OpenCV
+img: np.array = cv2.imread('./data/doraemon.png')
+img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB) / 255.0
+img = img + np.random.normal(loc=0.0, scale=0.1, size=img.shape)
+img = np.clip(img, 0.0, 1.0)
+
+# convert to torch tensor
+noisy_image: torch.tensor = kornia.image_to_tensor(img).squeeze()  # CxHxW
+noisy_image
```
Comment (ducha-aiki) [inline examples/total_variation_denoising.py]: remove

File: examples/total_variation_denoising.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# define the optimizer to optimize the 1 parameter of tv_denoiser
+optimizer = torch.optim.SGD(tv_denoiser.parameters(), lr=0.1, momentum=0.9)
+
+# run the optimization loop
+num_iters = 500
+for i in range(num_iters):
+    optimizer.zero_grad()
+    loss = tv_denoiser()
+    if i % 25 == 0:
+        print("Loss in iteration ", i, " of ", num_iters, ": ", loss.detach().numpy())
```
Comment (ducha-aiki) [inline examples/total_variation_denoising.py]: loss.detach().numpy() -> loss.item() 

Nicer output, no need of array

File: kornia/losses/total_variation.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if len(img_shape) == 3:
+        pixel_dif1 = img[:, 1:, :] - img[:, :-1, :]
+        pixel_dif2 = img[:, :, 1:] - img[:, :, :-1]
+        reduce_axes = (0, 1, 2)
+    elif len(img_shape) == 4:
+        pixel_dif1 = img[:, :, 1:, :] - img[:, :, :-1, :]
+        pixel_dif2 = img[:, :, :, 1:] - img[:, :, :, :-1]
+        reduce_axes = (1, 2, 3)
+    else:
+        raise ValueError("Expected input tensor to be of rank 3 or 4, but got " + str(len(img_shape)))
```
Comment (ducha-aiki) [inline kornia/losses/total_variation.py]: I believe, it is not "rank", but ndim. 

"The total number of contravariant and covariant indices of a tensor. The rank R of a tensor is independent of the number of dimensions N of the underlying space."

http://mathworld.wolfram.com/TensorRank.html

File: test/test_losses.py [hunk 5]
```diff
@@ -265,3 +267,58 @@ def test_kl_div_loss_2d(self, input, target, expected):
...
+    ])
+    def test_tv_on_3d(self, input):
+        actual = kornia.losses.total_variation(input)
+        input_np = input.numpy()
+        pixel_dif1 = input_np[:, 1:, :] - input_np[:, :-1, :]
+        pixel_dif2 = input_np[:, :, 1:] - input_np[:, :, :-1]
+        expected = torch.tensor(np.sum(np.abs(pixel_dif1)) + np.sum(np.abs(pixel_dif2)))
+        assert_allclose(actual, expected)
+
+    # Total variation for 3D tensors
```
Comment (ducha-aiki) [inline test/test_losses.py]: 4D

File: kornia/losses/total_variation.py (hunk 4 above)
Comment (edgarriba) [reply kornia/losses/total_variation.py]: This is not needed using ellipsis. We can handle (*,H,W)
```python
pixel_dif1 = img[..., 1:, :] - img[..., :-1, :]
pixel_dif2 = img[..., :, 1:] - img[..., :, :-1]
```

Comment (edgarriba) [reply kornia/losses/total_variation.py]: Besides I believe it can be done with a conv2d using a filter of [1, -1] including padding so that we can recover the full resolution image

Comment (edgarriba) [reply kornia/losses/total_variation.py]: Whatever is faster is fine to me

File: examples/total_variation_denoising.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+# define the optimizer to optimize the 1 parameter of tv_denoiser
+optimizer = torch.optim.SGD(tv_denoiser.parameters(), lr=0.1, momentum=0.9)
+
+# run the optimization loop
+num_iters = 500
+for i in range(num_iters):
+    optimizer.zero_grad()
+    loss = tv_denoiser()
+    if i%25 == 0:
+        print("Loss in iteration ", i, " of ", num_iters, ": ", loss.item())
```
Comment (dvd42) [inline examples/total_variation_denoising.py]: please use the `'{}'.format()` format for printing. Something like:
`print("Loss in iteration {} of {}: {:.3f}".format(i, num_iters, loss.item()))`
It's much cleaner

File: examples/total_variation_denoising.py (hunk 1 above)
Comment (ananyamukh6) [reply examples/total_variation_denoising.py]: Removed extra import

File: examples/total_variation_denoising.py (hunk 2 above)
Comment (ananyamukh6) [reply examples/total_variation_denoising.py]: Removed

File: examples/total_variation_denoising.py (hunk 3 above)
Comment (ananyamukh6) [reply examples/total_variation_denoising.py]: Thanks for the suggestion. Used `item`

File: test/test_losses.py (hunk 5 above)
Comment (ananyamukh6) [reply test/test_losses.py]: Fixed comment

File: examples/total_variation_denoising.py (hunk 6 above)
Comment (ananyamukh6) [reply examples/total_variation_denoising.py]: Good idea! Done.

File: test/test_losses.py [hunk 7]
```diff
@@ -265,3 +267,58 @@ def test_kl_div_loss_2d(self, input, target, expected):
...
+    @pytest.mark.parametrize('input', [
+        torch.rand(2, 3, 4, 5),
+        torch.rand(3, 1, 2, 3),
+    ])
+    def test_tv_on_4d(self, input):
+        actual = kornia.losses.total_variation(input)
+        input_np = input.numpy()
+        pixel_dif1 = input_np[:, :, 1:, :] - input_np[:, :, :-1, :]
+        pixel_dif2 = input_np[:, :, :, 1:] - input_np[:, :, :, :-1]
+        expected = torch.tensor(np.sum(np.abs(pixel_dif1), axis=(1, 2, 3)) + np.sum(np.abs(pixel_dif2), axis=(1, 2, 3)))
```
[the remark this answers is not in the document]
Comment (ananyamukh6) [reply test/test_losses.py]: The TotalVariation code does these ops in torch terms. This is a test, hence I am using numpy to compute the expected values. Using torch to compute the values expected of the torch function is slightly circular.

File: kornia/losses/total_variation.py (hunk 4 above)
Comment (ananyamukh6) [reply kornia/losses/total_variation.py]: @edgarriba 
I did think of using convolution. But the example for denoising I have ran in ~9 secs with conv (while with the current diff implementation it ran in ~2 secs). (ran on CPU, not GPU). The time diff was quite surprising to me!

here is the code for total variation implemented in terms of convolution. Let me know if you see any scope of optimizing the conv based implementation and I can try it:
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.conv_3d_1 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(1,2), bias=False)
        self.conv_3d_1.weight = torch.nn.Parameter(data=torch.tensor([[[[1.0,-1.0]]]]), requires_grad=False)
        self.conv_3d_2 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(2,1), bias=False)
        self.conv_3d_2.weight = torch.nn.Parameter(data=torch.tensor([[[[ 1.0],[-1.0]]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        #return total_variation(img)
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(self.conv_3d_1(channel).abs().sum(dim=(1, 2, 3)))
            to_be_added.append(self.conv_3d_2(channel).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total


def total_variation(img: torch.Tensor) -> torch.Tensor:
    return TotalVariation()(img)
```

**Comment by ananyamukh6** [comment]
Addressed comments by reviewers. Please re-review

Also I see codecov fail (red crosses on them). Can someone familiar with the repo please tell me if that is expected, and if not what I can do to correct it.

File: test/test_losses.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dvd42) [reply test/test_losses.py]: I see your point, however it would be nice if the tests would have the least amount of dependencies as possible. Image numpy changes how it handles axis in the future, then the tests will fail when the code is correct. Of course, it is extremely unlikely that numpy changes how it handles axis in the future but you see my point. It is nice that the code  uses only torch, for consistency sake.

One way you could do these tests is by providing an input for which you know the expected output see [this example](https://github.com/arraiyopensource/kornia/blob/d97f76fe77ab19744a7def88e5dcefc71ddbae0e/test/color/test_hsv.py#L14). 

P.S: Thank you for your pacience, I am aware I am being a bit picky with these changes.

[the remark this answers is not in the document]
Comment (bhack) [reply test/test_losses.py]: The TF version use a double approach for the tests:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py#L4190

[the remark this answers is not in the document]
Comment (dvd42) [reply test/test_losses.py]: A double approach would also be valid althought I dont know if we necessarily need/want it

File: kornia/losses/total_variation.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    See :class:`~kornia.losses.TotalVariation` for details.
+    """
+    if not torch.is_tensor(img):
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
+    img_shape = img.shape
+    if len(img_shape) == 3 or len(img_shape) == 4:
+        pixel_dif1 = img[..., 1:, :] - img[..., :-1, :]
+        pixel_dif2 = img[..., :, 1:] - img[..., :, :-1]
+        reduce_axes = (0, 1, 2) if len(img_shape) == 3 else (1, 2, 3)
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/losses/total_variation.py]: yes, plus ellipsis

File: kornia/losses/total_variation.py (hunk 4 above)
Comment (edgarriba) [reply kornia/losses/total_variation.py]: filter2D already implements that. Give it a shot and check the performance
https://kornia.readthedocs.io/en/latest/filters.html#kornia.filters.filter2D

File: examples/total_variation_denoising.py (hunk 6 above)
Comment (edgarriba) [reply examples/total_variation_denoising.py]: not for this PR, but I recently started using f-string format, which is really cool !
https://realpython.com/python-f-strings/

File: test/test_losses.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bhack) [reply test/test_losses.py]: See also this comment https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py#L4284 and the follow-up handmade test.

[the remark this answers is not in the document]
Comment (edgarriba) [reply test/test_losses.py]: vote for avoid using numpy

[the remark this answers is not in the document]
Comment (ananyamukh6) [reply test/test_losses.py]: Removed numpy dependency

File: kornia/losses/total_variation.py (hunk 4 above)
Comment (ananyamukh6) [reply kornia/losses/total_variation.py]: Tried `kornia.filter.filter2D`, but it is still slow. I think that its just conv in this case is slower than subtraction. I suggest we stick to the current difference based implementation

Note I tried on CPU
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.weight1 = torch.nn.Parameter(data=torch.tensor([[[1.0,-1.0]]]), requires_grad=False)
        self.weight2 = torch.nn.Parameter(data=torch.tensor([[[ 1.0],[-1.0]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(kornia.filters.filter2D(channel, self.weight1).abs().sum(dim=(1, 2, 3)))

            to_be_added.append(kornia.filters.filter2D(channel, self.weight2).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total
```

**Comment by ananyamukh6** [comment]
Addressed all comments. Please re-review.

I notice that CI is failing, but its failing at the "install dependencies" stage. Is CI down?

File: kornia/losses/total_variation.py (hunk 4 above)
Comment (edgarriba) [reply kornia/losses/total_variation.py]: Great

**Comment by ananyamukh6** [comment]
> LGTM. Awesome contribution. Will merge once we fix the CI build issue

Thanks!
CI is passing now

**Comment by edgarriba** [comment]
@ananyamukh6 you made it ! thanks for the contribution. Anything else in your mind to be implemented ?
